creusot/src/resolve.rs

```diff
@@ -1,205 +1,283 @@
+            .mark_inactive_variants_as_uninit()
             .into_engine(tcx, body)
             .iterate_to_fixpoint()
             .into_results_cursor(body);
 
         // MaybeLiveExceptDrop ignores `drop` for the purpose of resolve liveness... unclear that this can
         // be sound.
-        let local_live = MaybeLiveExceptDrop
+        let live = MaybeLiveExceptDrop::new(body, mdpe, tcx)
             .into_engine(tcx, body)
             .iterate_to_fixpoint()
             .into_results_cursor(body);
 
-        let borrows_out_of_scope =
-            calculate_borrows_out_of_scope_at_location(body, &regioncx, &borrow_set);
-
-        let borrows = Borrows::new(tcx, body, borrow_set.clone(), borrows_out_of_scope.clone())
+        let borrows = Borrows::new(tcx, body, &regioncx, borrow_set)
             .into_engine(tcx, body)
             .iterate_to_fixpoint()
             .into_results_cursor(body);
 
-        EagerResolver { local_live, local_init, local_uninit, borrows, borrow_set, body }
-    }
-
-    fn seek_to(&mut self, loc: ExtendedLocation) {
-        loc.seek_to(&mut self.local_live);
-        loc.seek_to(&mut self.local_init);
-        loc.seek_to(&mut self.local_uninit);
-        loc.seek_to(&mut self.borrows);
-    }
-
-    fn def_init_locals(&self) -> BitSet<Local> {
-        let init = self.local_init.get().clone();
-        let mut uninit: BitSet<_> = BitSet::new_empty(self.body.local_decls.len());
-        uninit.union(self.local_uninit.get());
-
-        let mut def_init = init;
-        def_init.subtract(&uninit);
-        def_init
+        EagerResolver { live, uninit, borrows, borrow_set, mdpe, body, tcx }
     }
 
-    fn live_locals(&self) -> BitSet<Local> {
-        let mut live: BitSet<_> = BitSet::new_empty(self.body.local_decls.len());
-        live.union(self.local_live.get());
-        live
-    }
-
-    fn dead_locals(&self) -> BitSet<Local> {
-        let live = self.live_locals();
-        let mut dead: BitSet<_> = BitSet::new_filled(live.domain_size());
-        dead.subtract(&live);
-        dead
-    }
-
-    fn frozen_locals(&self) -> BitSet<Local> {
-        let mut frozen: BitSet<_> = BitSet::new_empty(self.body.local_decls.len());
-        for bi in self.borrows.get().iter() {
-            let l = self.borrow_set[bi].borrowed_place.local;
-            frozen.insert(l);
+    fn frozen_of_borrows(&self, borrows: &BitSet<BorrowIndex>) -> BitSet<MovePathIndex> {
+        let mut frozen = self.empty_bitset();
+        for bi in borrows.iter() {
+            let place = self.borrow_set[bi].borrowed_place;
+            match self.move_data().rev_lookup.find(place.as_ref()) {
+                LookupResult::Exact(mp) => on_all_children_bits(&self.move_data(), mp, |mp| {
+                    frozen.insert(mp);
+                }),
+                LookupResult::Parent(mp) => drop(frozen.insert(mp.unwrap())),
+            };
         }
         frozen
     }
```

Comment:
You get a set with two elements, which correspond to the two borrowed move paths. Note that the parent move path will not be considered borrowed, and thus be resolved earlier (which will not be visible, because resolving the parent place without its children is a no-op).

Adding a comment.